// @similarweb/styles/dist/webpack/src/fonts.ts
export const $sansFontFamily = `'Roboto', Tahoma, sans-serif`;
export const $fontCondensed = `'Roboto', Tahoma, sans-serif`;
export const $fontCondensedLight = `'Roboto Light', Tahoma, sans-serif`;
export const $serifFontFamily = `Georgia, "Times New Roman", Times, serif !default`;
export const $robotoFontFamily = `"Roboto", sans-serif`;
export const $dmSansFontFamily = `'DM Sans', 'Roboto', Tahoma, sans-serif`;
export const $swBaseFontSize = 16;
export const $baseFontSize = 12;

export type FontVariant =
  | "h1"
  | "h2"
  | "h3"
  | "h4"
  | "h5"
  | "h6"
  | "report"
  | "subsectionTitle"
  | "primaryItem"
  | "sectionTitle"
  | "subtitle1"
  | "subtitle2"
  | "subTitle3"
  | "input1"
  | "input2"
  | "body1"
  | "body2"
  | "button"
  | "tab"
  | "listItem1"
  | "listItem3"
  | "listItem4"
  | "tooltip"
  | "value"
  | "listSectionTitle"
  | "captionTitle"
  | "caption"
  | "caption2"
  | "dataTable1"
  | "dataTable2"
  | "queryBar1"
  | "tag"
  | "chip1"
  | "chip1TopRow";

interface FontOptions {
  family: string;
  weight: string | number;
  size: string | number;
  lineHeight: string | number;
}

export const fontVariants: Record<FontVariant, FontOptions> = {
  h1: {
    family: $dmSansFontFamily,
    weight: 500,
    size: 48,
    lineHeight: 56,
  },
  h2: {
    family: $dmSansFontFamily,
    weight: 500,
    size: 34,
    lineHeight: 40,
  },
  h3: {
    family: $dmSansFontFamily,
    weight: 400,
    size: 32,
    lineHeight: 40,
  },
  h4: {
    family: $dmSansFontFamily,
    weight: 500,
    size: 24,
    lineHeight: 32,
  },
  h5: {
    family: $dmSansFontFamily,
    weight: 500,
    size: 20,
    lineHeight: 24,
  },
  h6: {
    family: $dmSansFontFamily,
    weight: 500,
    size: 18,
    lineHeight: 24,
  },
  subsectionTitle: {
    family: $dmSansFontFamily,
    weight: 700,
    size: 16,
    lineHeight: 20,
  },
  primaryItem: {
    family: $dmSansFontFamily,
    weight: 500,
    size: 14,
    lineHeight: 16,
  },
  sectionTitle: {
    family: $dmSansFontFamily,
    weight: 400,
    size: 12,
    lineHeight: 16,
  },
  report: {
    family: $dmSansFontFamily,
    weight: 400,
    size: 14,
    lineHeight: 16,
  },
  subtitle1: {
    family: $robotoFontFamily,
    weight: 500,
    size: 16,
    lineHeight: 20,
  },
  subtitle2: {
    family: $robotoFontFamily,
    weight: 500,
    size: 14,
    lineHeight: 16,
  },
  subTitle3: {
    family: $robotoFontFamily,
    weight: 700,
    size: 16,
    lineHeight: 20,
  },
  body1: {
    family: $robotoFontFamily,
    weight: 400,
    size: 16,
    lineHeight: 20,
  },
  body2: {
    family: $robotoFontFamily,
    weight: 400,
    size: 14,
    lineHeight: 20,
  },
  input1: {
    family: $robotoFontFamily,
    weight: 400,
    size: 24,
    lineHeight: 24,
  },
  input2: {
    family: $robotoFontFamily,
    weight: 400,
    size: 14,
    lineHeight: 16,
  },
  button: {
    family: $dmSansFontFamily,
    weight: 700,
    size: 14,
    lineHeight: 20,
  },
  tab: {
    family: $dmSansFontFamily,
    weight: 700,
    size: 14,
    lineHeight: 16,
  },
  listItem1: {
    family: $robotoFontFamily,
    weight: 500,
    size: 14,
    lineHeight: 20,
  },
  listItem3: {
    family: $robotoFontFamily,
    weight: 400,
    size: 12,
    lineHeight: 16,
  },
  listItem4: {
    family: $robotoFontFamily,
    weight: 700,
    size: 13,
    lineHeight: 20,
  },
  tooltip: {
    family: $robotoFontFamily,
    weight: 400,
    size: 13,
    lineHeight: 20,
  },
  value: {
    family: $robotoFontFamily,
    weight: 400,
    size: 13,
    lineHeight: 16,
  },
  listSectionTitle: {
    family: $robotoFontFamily,
    weight: 500,
    size: 12,
    lineHeight: 16,
  },
  captionTitle: {
    family: $robotoFontFamily,
    weight: 500,
    size: 12,
    lineHeight: 16,
  },
  caption: {
    family: $robotoFontFamily,
    weight: 400,
    size: 12,
    lineHeight: 20,
  },
  dataTable1: {
    family: $robotoFontFamily,
    weight: 500,
    size: 11,
    lineHeight: 16,
  },
  dataTable2: {
    family: $robotoFontFamily,
    weight: 400,
    size: 11,
    lineHeight: 16,
  },
  queryBar1: {
    family: $robotoFontFamily,
    weight: 400,
    size: 10,
    lineHeight: 12,
  },
  caption2: {
    family: $robotoFontFamily,
    weight: 500,
    size: 9,
    lineHeight: 16,
  },
  tag: {
    family: $robotoFontFamily,
    weight: 700,
    size: 9,
    lineHeight: 8,
  },
  chip1: {
    family: $robotoFontFamily,
    weight: 400,
    size: 13,
    lineHeight: 16,
  },
  chip1TopRow: {
    family: $robotoFontFamily,
    weight: 500,
    size: 13,
    lineHeight: 16,
  },
};
