// @similarweb/styles/dist/webpack/src/style-guide/colorsSets.ts
import {
  blue,
  bluegrey,
  midnight,
  mint,
  orange,
  red,
  pink,
  purple,
  carbon,
  green,
  sky,
  teal,
  solutions,
} from "./colorsPalettes";
import { ColorsGroup } from "./ColorsGroup";

export const brand = ColorsGroup("brand", {
  "midnight 500": midnight["500"],
  "bluegrey 100": bluegrey["100"],
  "blue 400": blue["400"],
  "orange 400": orange["400"],
  "mint 400": mint["400"],
});

export const a1 = ColorsGroup("a1", {
  "blue 400": blue["400"],
});

export const a2 = ColorsGroup("a2", {
  "blue 100": blue["100"],
  "blue 200": blue["200"],
  "sky 300": sky["300"],
  "blue 400": blue["400"],
  "blue 500": blue["500"],
});

export const b1 = ColorsGroup("b1", {
  "blue 400": blue["400"],
  "sky 300": sky["300"],
});
export const b2 = ColorsGroup("b2", {
  "midnight 400": midnight["400"],
  "midnight 300": midnight["300"],
});

export const b3 = ColorsGroup("b3", {
  "midnight 50": midnight["50"],
});
export const c = ColorsGroup("c", {
  "midnight 300": midnight["300"],
  "torquoise 300": orange["400"],
  "mint 400": mint["400"],
  SALES: solutions["SALES"],
  DIGITAL_RESEARCH: solutions["DIGITAL_RESEARCH"],
  DIGITAL_MARKETING: pink["600"],
  "purple 500": purple["500"],
  "INVESTORS 300": teal["700"],
  "red 400": red["500"],
  SHOPPER: solutions["SHOPPER"],
  others: carbon["50"],
});

export const verdics = ColorsGroup("verdics", {
  "red s100": red["s100"],
  "green s100": green["s100"],
});

export const diverging = ColorsGroup("diverging", {
  "red 400": red["400"],
  "sky 200": sky["200"],
  "blue 400": blue["400"],
});

export const divergingSequential = ColorsGroup("diverging sequential", {
  "P5": blue["500"],
  "P4": blue["400"],
  "P3": blue["350"],
  "P2": blue["300"],
  "P1": blue["200"],
  "N1": red["200"],
  "N2": red["300"],
  "N3": red["350"],
  "N4": red["400"],
  "N5": red["500"],
});
